lib/linked_list.rb

```diff
@@ -18,129 +19,255 @@ def initialize
 
     # method to add a new node with the specific data value in the linked list
     # insert the new node at the beginning of the linked list
-    # Time Complexity:  
-    # Space Complexity
+    # Time Complexity: O(1) (constant), because the length of the list doesn't matter
+    # Space Complexity: O(n), where n is the length of the list
     def add_first(value)
-      raise NotImplementedError
+      new_node = Node.new(value, nil)
+      new_node.next = @head
+      @head = new_node
+
+      return new_node.data
     end
 
+    # Additional Exercises
+        # returns the value in the first node
+        # returns nil if the list is empty
+        # Time Complexity: O(1)
+        # Space Complexity: O(1)
+        def get_first
+          return nil if @head.nil?
+
+          return @head.data
+        end
+
     # method to find if the linked list contains a node with specified value
     # returns true if found, false otherwise
-    # Time Complexity:  
-    # Space Complexity
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def search(value)
-      raise NotImplementedError
+      node = @head
+          return false if !node.next

```

Comment:
What if @head is nil?
What if the value is in the head and the list only has 1 element?